### PR TITLE
SRV-304 Add currency feature

### PR DESCRIPTION
**Pull request by PawelPodkalicki**

### CODE REVIEW

File: database/migrations/2019_04_19_123103_change_event_logs_add_currency.php [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+ */
+
+use Adshares\Adserver\Facades\DB;
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class ChangeEventLogsAddCurrency extends Migration
+{
+    private const EXCHANGE_RATE_VALUE = '0.05';
```
Comment (PawelPodkalicki) [inline database/migrations/2019_04_19_123103_change_event_logs_add_currency.php]: Should be updated to current exchange rate.

File: app/Client/AdsOperatorExchangeRateRepository.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+use Adshares\Adserver\Utilities\DateUtils;
+use Adshares\Common\Application\Dto\ExchangeRate;
+use Adshares\Common\Application\Service\Exception\ExchangeRateNotAvailableException;
+use Adshares\Common\Application\Service\ExchangeRateRepository;
+use DateTime;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\RequestException;
+
+final class AdsOperatorExchangeRateRepository implements ExchangeRateRepository
```
Comment (c3zi) [inline app/Client/AdsOperatorExchangeRateRepository.php]: Why not ```GuzzleAdsOperatorClient```?

File: app/Console/Commands/DemandBlockRequiredAmount.php [hunk 3]
```diff
@@ -35,28 +38,41 @@ class DemandBlockRequiredAmount extends Command
...
+        $this->exchangeRateReader = $exchangeRateReader;
+
+        parent::__construct();
+    }
+
     public function handle(): void
     {
         $this->info('Start command '.$this->signature);
 
+        $exchangeRate = $this->exchangeRateReader->fetchExchangeRate(new DateTime());
```
Comment (c3zi) [inline app/Console/Commands/DemandBlockRequiredAmount.php]: I'd suggest setting ```new DateTime()``` value by default in ```fetchExchangeRate``` method. 
It occurs several times in the code.

File: app/Models/Campaign.php [hunk 4]
```diff
@@ -252,53 +253,53 @@ public function getBasicInformationAttribute(): array
...
-        if ($status === self::STATUS_ACTIVE) {
-            $amount += $budgetForCurrentDateTime;
-        } elseif ($this->status === self::STATUS_ACTIVE) {
-            $amount -= $budgetForCurrentDateTime;
+        if (null === $exchangeRate) {
+            Log::error('ExchangeRate is not available');
+            throw new InvalidArgumentException('ExchangeRate is not available');
         }
 
+        $totalBudget = self::fetchRequiredBudgetForAllCampaignsInCurrentPeriod() + $budget;
```
Comment (c3zi) [inline app/Models/Campaign.php:286]: Does it mean that we didn't include the needed budget from the current hour before?

**Review by c3zi — CHANGES_REQUESTED** [review]

File: app/Client/AdsOperatorExchangeRateRepository.php (hunk 2 above)
Comment (PawelPodkalicki) [reply app/Client/AdsOperatorExchangeRateRepository.php]: I'll rename.

File: app/Console/Commands/DemandBlockRequiredAmount.php (hunk 3 above)
Comment (PawelPodkalicki) [reply app/Console/Commands/DemandBlockRequiredAmount.php]: Good point. I'll change.

File: app/Models/Campaign.php (hunk 4 above)
Comment (PawelPodkalicki) [reply app/Models/Campaign.php:286]: Blockade update is done by releasing previous blockade.

**Review by c3zi — DISMISSED** [review]

**Review by c3zi — APPROVED** [review]

**Pull request merged**